mining: Improve block template generator test coverage

santi-h opened the pull request:
### All Submissions:

* [ ] Submission follows the [Code Contribution Guidelines](../blob/master/docs/code_contribution_guidelines.md)
* [ ] There are not any other open [Pull Requests](../pulls) for the same update/change
* [ ] Commit messages are formatted according to [Model Git Commit Messages](../blob/master/docs/code_contribution_guidelines.md#44-model-git-commit-messages)
* [ ] All changes are compliant with the latest version of Go and the one prior to it
* [ ] The code being submitted is commented according to the [Code Documentation and Commenting](../blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) section of the Code Contribution Guidelines
* [ ] Any new logging statements use an appropriate subsystem and logging level
* [ ] Code has been formatted with go fmt
* [ ] Running go test does not fail any tests or report any vet issues
* [ ] Running golint does not report any new issues that did not already exist

### New Feature Submissions:

* [ ] Code is accompanied by tests which exercise both the positive and negative (error paths) conditions (if applicable)

### Bug Fixes:

* [ ] Code is accompanied by new tests which trigger the bug being fixed to prevent regressions

### Changes to Core Features:

* [ ] An explanation of what the changes do and why they should be included is provided
* [ ] Code is accompanied by updates to tests and/or new tests for the core changes (if applicable)

rstaudt2 commented on internal/mining/mining_test.go:
```diff
@@ -59,13 +59,62 @@ func TestNewBlockTemplateBasicErrorScenarios(t *testing.T) {
+	// Test that the error is triggered
+	_, err = harness.generator.NewBlockTemplate(address)
+	if err == nil {
+		t.Fatalf("expected error retrieving tip generation -- got nil")
```
Can use the non-formatted version if formatting unneeded. 
```suggestion
		t.Fatal("expected error retrieving tip generation -- got nil")
```

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -79,6 +79,54 @@ type fakeChain struct {
+func NewFakeChain(chainParams *chaincfg.Params, heightOfBestState int64) *fakeChain {
+	if heightOfBestState < int64(chainParams.CoinbaseMaturity) {
+		heightOfBestState = int64(chainParams.CoinbaseMaturity)
+	}
```
I'll comment on the overall design too, so this may not matter, but I don't see any reason why you'd not allow setting a best height less than coinbase maturity.

santi-h replied:
I was just trying to keep the current functionality as is. Right now it's impossible for the best height to be less than the coinbase maturity because of [this](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/internal/mining/mining_harness_test.go#L1499). Well technically not impossible because `curHeight` is a signed int and could be negative (not sure how realistic this would be), but right now it's [getting the value from `chain.bestState.Height` when `chain.bestState` is not yet initialized](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/internal/mining/mining_harness_test.go#L1480) so it's zero anyways.

Do you happen to know why it's doing `int64(chainParams.CoinbaseMaturity) + curHeight` and not just setting it to `curHeight`? Maybe I can just remove this dependency between the two and see if some test breaks. I just thought there was a reason behind it.

rstaudt2 replied:
> Do you happen to know why it's doing `int64(chainParams.CoinbaseMaturity) + curHeight` and not just setting it to `curHeight`? Maybe I can just remove this dependency between the two and see if some test breaks. I just thought there was a reason behind it.

The only reason it's setting it to `int64(chainParams.CoinbaseMaturity) + curHeight` because it's creating an output from a coinbase and a coinbase output is not spendable until `chainParams.CoinbaseMaturity`, and the intention is to return spendable outputs for the tests to start with.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -79,6 +79,54 @@ type fakeChain struct {
+// newMiningHarness.
+// If the value of heightOfBestState is less than chainParams.CoinbaseMaturity, it will be
+// set to chainParams.CoinbaseMaturity
+func NewFakeChain(chainParams *chaincfg.Params, heightOfBestState int64) *fakeChain {
```
It's a bit odd to take `heightOfBestState` as a parameter but don't allow setting anything else on the chain (including the other fields of best state).  It seems like it'd either be better to take a config type that holds everything or just let the caller set those fields on the chain as needed afterward (basically how it currently is).

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -79,6 +79,54 @@ type fakeChain struct {
+		fetchUtxoEntryErr:                  nil,
+		fetchUtxoViewErr:                   nil,
+		fetchUtxoViewParentTemplateErr:     nil,
+		forceHeadReorganizationErr:         nil,
```
You may just be doing this to be explicit, but most of these you are setting to the default value (`0`, `nil`, `false`, etc.), which is unnecessary.

santi-h replied:
Yup, I'll remove.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -1142,6 +1190,18 @@ func (m *miningHarness) CreateSignedTx(inputs []spendableOutput, numOutputs uint
 	return dcrutil.NewTx(tx), nil
 }
 
+// Returns a new chaincfg.Params to be used for testing with the mining harness
```
Function commends should always start with the name of the function.

rstaudt2 commented on internal/mining/mining_harness_test.go:
```diff
@@ -1317,7 +1377,16 @@ func (m *miningHarness) AddFakeUTXO(tx *dcrutil.Tx, blockHeight int64, blockInde
+	// Set parameters to their default values if nothing passed
+	if chainParams == nil {
+		chainParams = newChainParamsForMiningHarness()
+	}
```
An alternative pattern that I like a bit better than passing `nil` for this is defining a separate function.  For example, you could define `newMiningHarnessWithChain` that takes a chain param, and this `newMiningHarness` function could just create the chain and then call `newMiningHarnessWithChain` internally.  Then the caller can decide if they want to use `newMiningHarness` or `newMiningHarnessWithChain`, which is more obvious than knowing you could pass `nil` for the chain.

santi-h replied:
Ya good idea, I'll change it to follow that pattern.

rstaudt2 commented on internal/mining/mining_test.go:
```diff
@@ -59,13 +59,62 @@ func TestNewBlockTemplateBasicErrorScenarios(t *testing.T) {
 	harness.chain.isTreasuryAgendaActiveErr = nil
 }
 
+// TestTipGenerationError tests the TipGeneration error scenario
```
Very minor, but usually I'd avoid using names like `TipGeneration` and instead just put "tip generation" unless I am describe something about a specific type or function.  The benefit is that if the type/function/interface/etc. changes the comment is usually still accurate.

rstaudt2 commented on internal/mining/mining_test.go:
```diff
@@ -59,13 +59,62 @@ func TestNewBlockTemplateBasicErrorScenarios(t *testing.T) {
+	} else if errorValue.Description != "error retrieving tip generation" {
+		t.Fatalf("expected Description to be '%v' -- got '%v'", "error retrieving tip generation",
+			errorValue.Description)
+	}
```
You should probably just use `errors.Is` here.  I'd recommend taking a look [at this](https://github.com/decred/dcrd/issues/2181) and how `errors.Is` is used for other tests.

rstaudt2 commented on internal/mining/mining_test.go:
```diff
@@ -59,13 +59,62 @@ func TestNewBlockTemplateBasicErrorScenarios(t *testing.T) {
+	// triggered
+	chain = NewFakeChain(chainParams, chainParams.StakeValidationHeight-2)
+	chain.tipGenerationErr = tipGenerationErr
+	harness, _, err = newMiningHarness(chainParams, chain)
```
I don't think there is a need to recreate the chain and harness.  Should be able to just update mocked state.

santi-h replied:
See comment [here](https://github.com/santi-h/dcrd/pull/1#issuecomment-968505662) for why I was doing this. I can update.

rstaudt2 reviewed (commented):
Looks like you are making some good progress on working through the mining code and adding more test coverage.  Nice!  A few comments on the overall design:
- The harness is intended to create everything that is needed to start from a pretty blank slate.  Is there any advantage to creating the chain separately vs. just creating the harness and then changing the chain state as needed for the tests afterward like the existing tests do?  This would simplify things quite a bit.
- I see you added a separate test case for the tip generation error.  When you get further into it, and more coverage is added, it's probably going to be very cumbersome to have a separate test case for every small error case.  I started doing something a bit more generic a while ago (see [this temp commit](https://github.com/rstaudt2/dcrd/commit/f745ce72a44e88879ecc4f04d661fc36919edc02)) that allows each test scenario to specify a munge function which allows that scenario to modify things on the harness or underlying chain.  It may be worth seeing if something like that makes it easier, you'll probably get a better and better idea of the best way to do it the more test cases you add.

Also, a few notes on some other code contribution guidelines.  You may already be aware and be planning on cleaning them up later, but FYI:
- First line of the commit needs to be <= `50` chars and should end with a `.`
- Other lines in the commit need to be <= `72` chars
- Code lines should not exceed `80` chars
  - If you are using vscode, you can use rulers to show where the line length is at:
  ```
      "editor.rulers": [
        72,
        80
      ],
  ```
- All comments should end with a `.`

santi-h commented:
>  Is there any advantage to creating the chain separately vs. just creating the harness and then changing the chain state as needed for the tests afterward like the existing tests do?

@rstaudt2 I didn't want to make changes to the chain state after the harness is created because I could very easily leave the harness in an inconsistent state. `newMiningHarness` uses the values from the `chain` to create other objects. For example it [uses `chain.isTreasuryAgendaActive` and `chain.bestState.Height` in order to `AddFakeUTXO`](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/internal/mining/mining_harness_test.go#L1485-L1486). If I change `chain.isTreasuryAgendaActive` or `chain.bestState.Height` after the harness was created then it would leave it in an inconsistent state unless I re-run `AddFakeUTXO` manually.

But I'm probably being a bit paranoid and it's not a big deal if the harness is in an inconsistent state. Plus it does make writing tests easier. For example [all these lines](https://github.com/santi-h/dcrd/pull/1/files#diff-c4bc044a45d389ce82fd8eadfc7a3ac5e3b1a576723029182f8c1f9a45a7cf9bR98-R103) will just become a one liner:
```golang
chain.bestState.Height = chainParams.StakeValidationHeight - 2
```

> I see you added a separate test case for the tip generation error. When you get further into it, and more coverage is added, it's probably going to be very cumbersome to have a separate test case for every small error case.

Ya I was thinking about that too. My plan is to start doing something dumb like this since I don't really know the code that well yet, and as I write more and more tests for `NewBlockTemplate` identify patterns to make writing them easier and refactor as I go. I will probably do something like what you did there with the anonymous structs. This was kind of a premature review request to make sure I'm not doing something completely crazy.

> Also, a few notes on some other code contribution guidelines. You may already be aware and be planning on cleaning them up later

Yeah that exactly what I was planning on doing. I appreciate you pointing those out though since it will save me time!

rstaudt2 commented:
> @rstaudt2 I didn't want to make changes to the chain state after the harness is created because I could very easily leave the harness in an inconsistent state. `newMiningHarness` uses the values from the `chain` to create other objects. For example it [uses `chain.isTreasuryAgendaActive` and `chain.bestState.Height` in order to `AddFakeUTXO`](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/internal/mining/mining_harness_test.go#L1485-L1486). If I change `chain.isTreasuryAgendaActive` or `chain.bestState.Height` after the harness was created then it would leave it in an inconsistent state unless I re-run `AddFakeUTXO` manually.

It shouldn't be an issue to change values after the harness is created.  What the harness is doing is just starting the chain at a point where you have some outputs to spend.  After that point, you can modify the underlying chain however you want to simulate things like more blocks/transactions/utxos being added or error scenarios. 

You are correct that it is easy to leave the chain in an unrealistic/inconsistent state, though.  For example, in reality you should be adding real blocks with corresponding transactions, utxos, and updating all values of the best state as well as several other things.  An alternative to this is to [use the chaingen harness](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/blockchain/common_test.go#L658-L677) which creates a real underlying chain.  You can see an example test I added recently for the consensus changes I was making [here](https://github.com/decred/dcrd/blob/ac76e08adfae210682095cffffec58fc90419b6f/blockchain/validate_test.go#L1768).  The downside to using that is it is harder to mock certain scenarios, which is why I started with a fake chain for this test suite, but I wouldn't be opposed to using that either.

santi-h committed:
mining: Add basic error test cases for NewBlockTemplate.

Adds test cases to NewBlockTemplate to make sure the following error
scenarios are covered:

- Error retrieving tip generation.
- Error retrieving auto revocations agenda.
- Error retrieving top block during handleTooFewVoters.
